File: sdk/python/symbolchain/symbol/Merkle.py
import hashlib
from binascii import hexlify
from collections import namedtuple
from enum import Enum
from functools import reduce

from ..BufferReader import BufferReader
from ..CryptoTypes import Hash256

MerklePart = namedtuple('MerklePart', ['hash', 'is_left'])
PatriciaTreePath = namedtuple('PatriciaTreePath', ['path', 'size'])

# region MerkleHashBuilder


class MerkleHashBuilder:
	"""Builder for creating a merkle hash."""

	def __init__(self):
		"""Creates a merkle hash builder."""
		self.hashes = []

	def update(self, component_hash):
		"""Adds a hash to the merkle hash."""
		self.hashes.append(component_hash.bytes)

	def final(self):
		"""Calculates the merkle hash."""
		if not self.hashes:
			return Hash256.zero()

		num_remaining_hashes = len(self.hashes)
		while num_remaining_hashes > 1:
			i = 0
			while i < num_remaining_hashes:
				hasher = hashlib.sha3_256()
				hasher.update(self.hashes[i])

				if i + 1 < num_remaining_hashes:
					hasher.update(self.hashes[i + 1])
				else:
					# if there is an odd number of hashes, duplicate the last one
					hasher.update(self.hashes[i])
					num_remaining_hashes += 1

				self.hashes[i // 2] = hasher.digest()
				i += 2

			num_remaining_hashes //= 2

		return Hash256(self.hashes[0])

# endregion


# region prove_merkle

def prove_merkle(leaf_hash, merkle_path, root_hash):
	"""
	Proves a merkle hash.
	Merkle *hash chain* path is ordered from leaf to root, where each element is MerklePart.
	"""

	def calculate_next_hash(working_hash, merkle_part):
		hasher = hashlib.sha3_256()
		if merkle_part.is_left:
			hasher.update(merkle_part.hash.bytes)
			hasher.update(working_hash.bytes)
		else:
			hasher.update(working_hash.bytes)
			hasher.update(merkle_part.hash.bytes)

		return Hash256(hasher.digest())

	computed_root_hash = reduce(calculate_next_hash, merkle_path, leaf_hash)
	return root_hash == computed_root_hash

# endregion


# region LeafNode / BranchNode

def _get_nibble_at(path, index):
	byte = path.path[index // 2]
	return byte & 0xF if 1 == index % 2 else byte >> 4


def _encode_path(path, is_leaf):
	i = 0
	buffer = [0x20 if is_leaf else 0]
	if 1 == path.size % 2:
		buffer[0] |= 0x10 | _get_nibble_at(path, 0)
		i += 1

	while i < path.size:
		buffer.append((_get_nibble_at(path, i) << 4) + _get_nibble_at(path, i + 1))
		i += 2

	return bytes(buffer)


class TreeNode:
	"""Node in a compact patricia tree."""

	def __init__(self, path):
		self.path = path

	@property
	def hex_path(self):
		"""Gets hex representation of path."""

		return hexlify(self.path.path).decode('utf8').upper()[:self.path.size]


class LeafNode(TreeNode):
	"""Leaf node in a compact patricia tree."""

	def __init__(self, path, value):
		super().__init__(path)
		self.value = value

	def calculate_hash(self):
		"""Calculates node hash."""

		hasher = hashlib.sha3_256()
		hasher.update(_encode_path(self.path, True))
		hasher.update(self.value.bytes)
		return Hash256(hasher.digest())


class BranchNode(TreeNode):
	"""Branch node in a compact patricia tree."""

	def __init__(self, path, links):
		super().__init__(path)
		self.links = links

	def calculate_hash(self):
		"""Calculates node hash."""

		hasher = hashlib.sha3_256()
		hasher.update(_encode_path(self.path, False))
		for link in self.links:
			hasher.update((link if link else Hash256.zero()).bytes)

		return Hash256(hasher.digest())

# endregion


# region deserialize_patricia_tree_nodes

def _deserialize_path(reader):
	num_nibbles = reader.read_int(1)
	num_bytes = (num_nibbles + 1) // 2
	return PatriciaTreePath(reader.read_bytes(num_bytes), num_nibbles)


def _deserialize_leaf(reader):
	path = _deserialize_path(reader)
	value = Hash256(reader.read_bytes(Hash256.SIZE))
	return LeafNode(path, value)


def _deserialize_branch(reader):
	path = _deserialize_path(reader)

	links_mask = reader.read_int(2)
	links = [None] * 16
	for index, _ in enumerate(links):
		if links_mask & (2 ** index):
			links[index] = Hash256(reader.read_bytes(Hash256.SIZE))

	return BranchNode(path, links)


def deserialize_patricia_tree_nodes(buffer):
	reader = BufferReader(buffer)
	nodes = []
	while not reader.eof:
		node_marker = reader.read_int(1)

		if 0xFF == node_marker:
			nodes.append(_deserialize_leaf(reader))
		elif 0x00 == node_marker:
			nodes.append(_deserialize_branch(reader))
		else:
			raise ValueError(f'invalid marker of a serialized node ({node_marker})')

	return nodes

# endregion


# region prove_patricia_merkle

class PatriciaMerkleProofResult(Enum):
	"""Possible results of a patricia merkle proof."""

	VALID_POSITIVE = 0x0001  # proof is valid (positive)
	VALID_NEGATIVE = 0x0002  # proof is valid (negative)

	INCONCLUSIVE = 0x4001  # negative proof is inconclusive

	STATE_HASH_DOES_NOT_MATCH_ROOTS = 0x8001  # state hash cannot be derived from subcache merkle roots
	UNANCHORED_PATH_TREE = 0x8002  # root of the path tree being proven is not a subcache merkle root
	LEAF_VALUE_MISMATCH = 0x8003  # leaf value does not match expected value
	UNLINKED_NODE = 0x8004  # provided merkle hash contains an unlinked node
	PATH_MISMATCH = 0x8005  # actual merkle path does not match encoded key


def _check_state_hash(state_hash, subcache_merkle_roots):
	hasher = hashlib.sha3_256()
	for root in subcache_merkle_roots:
		hasher.update(root.bytes)

	return state_hash == Hash256(hasher.digest())


def prove_patricia_merkle(encoded_key, value_to_test, merkle_path, state_hash, subcache_merkle_roots):
	"""
	Proves a patricia merkle hash.
	Merkle *node* path is ordered from root to leaf, where each element is either BranchNode or LeafNode.
	"""

	# pylint: disable=too-many-return-statements

	if not _check_state_hash(state_hash, subcache_merkle_roots):
		return PatriciaMerkleProofResult.STATE_HASH_DOES_NOT_MATCH_ROOTS

	if merkle_path[0].calculate_hash() not in subcache_merkle_roots:
		return PatriciaMerkleProofResult.UNANCHORED_PATH_TREE

	# positive proof must end with a leaf
	is_positive_proof = hasattr(merkle_path[-1], 'value')
	if is_positive_proof:
		if value_to_test != merkle_path[-1].value:
			return PatriciaMerkleProofResult.LEAF_VALUE_MISMATCH

	child_hash = None
	actual_path = ''
	for node in reversed(merkle_path):
		node_hash = node.calculate_hash()
		formatted_link_index = ''
		if child_hash:
			if child_hash not in node.links:
				return PatriciaMerkleProofResult.UNLINKED_NODE

			formatted_link_index = f'{node.links.index(child_hash):01X}'

		child_hash = node_hash
		actual_path = f'{formatted_link_index}{node.hex_path}{actual_path}'

	if is_positive_proof:
		# for positive proof, expected and calculated paths must match exactly
		return PatriciaMerkleProofResult.PATH_MISMATCH if actual_path != str(encoded_key) else PatriciaMerkleProofResult.VALID_POSITIVE

	# for negative proof, expected path must start with calculated path and next nibble must be a dead end
	if not str(encoded_key).startswith(actual_path):
		return PatriciaMerkleProofResult.PATH_MISMATCH

	next_nibble = _get_nibble_at(PatriciaTreePath(encoded_key.bytes, 2 * len(encoded_key.bytes)), len(actual_path))
	next_node = merkle_path[-1].links[next_nibble]
	return PatriciaMerkleProofResult.INCONCLUSIVE if next_node is not None else PatriciaMerkleProofResult.VALID_NEGATIVE

# endregion

File: sdk/javascript/src/symbol/merkle.js
import { Hash256 } from '../CryptoTypes.js';
import { deepCompare } from '../utils/arrayHelpers.js';
import { uint8ToHex } from '../utils/converter.js';
import { sha3_256 } from '@noble/hashes/sha3';

// region MerkleHashBuilder

/**
 * Builder for creating a merkle hash.
 */
export class MerkleHashBuilder {
	/**
	 * Creates a merkle hash builder.
	 */
	constructor() {
		/**
		 * @private
		 */
		this._hashes = [];
	}

	/**
	 * Adds a hash to the merkle hash.
	 * @param {Hash256} componentHash Hash to add.
	 */
	update(componentHash) {
		this._hashes.push(componentHash.bytes);
	}

	/**
	 * Calculates the merkle hash.
	 * @returns {Hash256} Merkle hash.
	 */
	final() {
		if (0 === this._hashes.length)
			return Hash256.zero();

		let numRemainingHashes = this._hashes.length;
		while (1 < numRemainingHashes) {
			let i = 0;
			while (i < numRemainingHashes) {
				const hasher = sha3_256.create();
				hasher.update(this._hashes[i]);

				if (i + 1 < numRemainingHashes) {
					hasher.update(this._hashes[i + 1]);
				} else {
					// if there is an odd number of hashes, duplicate the last one
					hasher.update(this._hashes[i]);
					numRemainingHashes += 1;
				}

				this._hashes[Math.floor(i / 2)] = hasher.digest();
				i += 2;
			}

			numRemainingHashes = Math.floor(numRemainingHashes / 2);
		}

		return new Hash256(this._hashes[0]);
	}
}

// endregion

// region proveMerkle

/**
 * Proves a merkle hash.
 * @param {Hash256} leafHash Leaf hash to prove.
 * @param {Array<MerklePart>} merklePath Merkle *hash chain* path from leaf to root.
 * @param {Hash256} rootHash Root hash of the merkle tree.
 * @returns {boolean} \c true if leaf hash is connected to root hash; false otherwise.
 */
const proveMerkle = (leafHash, merklePath, rootHash) => {
	const computedRootHash = merklePath.reduce((workingHash, merklePart) => {
		const hasher = sha3_256.create();
		if (merklePart.isLeft) {
			hasher.update(merklePart.hash.bytes);
			hasher.update(workingHash.bytes);
		} else {
			hasher.update(workingHash.bytes);
			hasher.update(merklePart.hash.bytes);
		}

		return new Hash256(hasher.digest());
	}, leafHash);

	return 0 === deepCompare(rootHash.bytes, computedRootHash.bytes);
};

// endregion

// region LeafNode / BranchNode

const getNibbleAt = (path, index) => {
	const byte = path.path[Math.floor(index / 2)];
	return 1 === index % 2 ? byte & 0xF : byte >>> 4;
};

const encodePath = (path, isLeaf) => {
	let i = 0;
	const buffer = new Uint8Array(1 + Math.floor(path.size / 2));
	buffer[0] = isLeaf ? 0x20 : 0;
	if (1 === path.size % 2) {
		buffer[0] |= 0x10 | getNibbleAt(path, 0);
		++i;
	}

	while (i < path.size) {
		buffer[1 + Math.floor(i / 2)] = (getNibbleAt(path, i) << 4) + (getNibbleAt(path, i + 1));
		i += 2;
	}

	return buffer;
};

/**
 *  Node in a compact Patricia tree.
 */
export class TreeNode {
	/**
	 * Creates a tree node.
	 * @param {PatriciaTreePath} path Node path.
	 */
	constructor(path) {
		/**
		 * Node path.
		 * @type PatriciaTreePath
		 */
		this.path = path;
	}

	/**
	 * Gets hex representation of path.
	 * @returns {string} Hex representation of path.
	 */
	get hexPath() {
		return uint8ToHex(this.path.path).substring(0, this.path.size);
	}

	/**
	 * Calculates node hash.
	 * @returns {Hash256} Hash of the node.
	 */
	calculateHash() { // eslint-disable-line class-methods-use-this
		return Hash256.zero();
	}
}

/**
 *  Leaf node in a compact Patricia tree.
 */
export class LeafNode extends TreeNode {
	/**
	 * Creates a leaf node.
	 * @param {PatriciaTreePath} path Leaf path.
	 * @param {Hash256} value Leaf value.
	 */
	constructor(path, value) {
		super(path);

		/**
		 * Leaf value.
		 * @type {Hash256}
		 */
		this.value = value;
	}

	/**
	 * Calculates node hash.
	 * @override
	 * @returns {Hash256} Hash of the node.
	 */
	calculateHash() {
		const hasher = sha3_256.create();
		hasher.update(encodePath(this.path, true));
		hasher.update(this.value.bytes);
		return new Hash256(hasher.digest());
	}
}

/**
 *  Branch node in a compact Patricia tree.
 */
export class BranchNode extends TreeNode {
	/**
	 * Creates a branch node.
	 * @param {PatriciaTreePath} path Branch path.
	 * @param {Array<Hash256>} links Branch links.
	 */
	constructor(path, links) {
		super(path);

		/**
		 * Branch links.
		 * @type Array<Hash256>
		 */
		this.links = links;
	}

	/**
	 * Calculates node hash.
	 * @override
	 * @returns {Hash256} Hash of the node.
	 */
	calculateHash() {
		const hasher = sha3_256.create();
		hasher.update(encodePath(this.path, false));
		this.links.forEach(link => {
			hasher.update((undefined === link ? Hash256.zero() : link).bytes);
		});

		return new Hash256(hasher.digest());
	}
}

// endregion

// region deserializePatriciaTreeNodes

class BufferReader {
	constructor(buffer) {
		this.view = new DataView(buffer);
		this.offset = 0;
	}

	get eof() {
		return this.offset === this.view.byteLength;
	}

	readByte() {
		const result = this.view.getUint8(this.offset);
		++this.offset;
		return result;
	}

	readShort() {
		const result = this.view.getUint16(this.offset, true);
		this.offset += 2;
		return result;
	}

	readBytes(count) {
		const result = new Uint8Array(this.view.buffer, this.offset, count);
		this.offset += count;
		return result;
	}
}

const deserializePath = reader => {
	const numNibbles = reader.readByte();
	const numBytes = Math.floor((numNibbles + 1) / 2);
	return { path: reader.readBytes(numBytes), size: numNibbles };
};

const deserializeLeaf = reader => {
	const path = deserializePath(reader);
	const value = new Hash256(reader.readBytes(Hash256.SIZE));
	return new LeafNode(path, value);
};

const deserializeBranch = reader => {
	const path = deserializePath(reader);

	const linksMask = reader.readShort();
	const links = new Array(16);
	for (let i = 0; i < links.length; ++i)
		links[i] = linksMask & (2 ** i) ? new Hash256(reader.readBytes(Hash256.SIZE)) : undefined;

	return new BranchNode(path, links);
};

/**
 * Deserializes a buffer containing patricia tree nodes.
 * @param {Uint8Array} buffer Buffer containing serialized patricia tree nodes.
 * @returns {Array<TreeNode>} Deserialized patricia tree nodes.
 */
const deserializePatriciaTreeNodes = buffer => {
	const reader = new BufferReader(buffer.buffer);
	const nodes = [];
	while (!reader.eof) {
		const nodeMarker = reader.readByte();

		switch (nodeMarker) {
		case 0xFF:
			nodes.push(deserializeLeaf(reader));
			break;

		case 0x00:
			nodes.push(deserializeBranch(reader));
			break;

		default:
			throw new Error(`invalid marker of a serialized node (${nodeMarker})`);
		}
	}

	return nodes;
};

// endregion

// region provePatriciaMerkle

/**
 * Possible results of a patricia merkle proof.
 */
export class PatriciaMerkleProofResult {
	/**
	 * Proof is valid (positive).
	 * @type number
	 */
	static VALID_POSITIVE = 0x0001;

	/**
	 * Proof is valid (negative).
	 * @type number
	 */
	static VALID_NEGATIVE = 0x0002;

	/**
	 * Negative proof is inconclusive.
	 * @type number
	 */
	static INCONCLUSIVE = 0x4001;

	/**
	 * State hash cannot be derived from subcache merkle roots.
	 * @type number
	 */
	static STATE_HASH_DOES_NOT_MATCH_ROOTS = 0x8001;

	/**
	 * Root of the path tree being proven is not a subcache merkle root.
	 * @type number
	 */
	static UNANCHORED_PATH_TREE = 0x8002;

	/**
	 * Leaf value does not match expected value.
	 * @type number
	 */
	static LEAF_VALUE_MISMATCH = 0x8003;

	/**
	 * Provided merkle hash contains an unlinked node.
	 * @type number
	 */
	static UNLINKED_NODE = 0x8004;

	/**
	 * Actual merkle path does not match encoded key.
	 * @type number
	 */
	static PATH_MISMATCH = 0x8005;
}

const checkStateHash = (stateHash, subcacheMerkleRoots) => {
	const hasher = sha3_256.create();
	subcacheMerkleRoots.forEach(root => {
		hasher.update(root.bytes);
	});

	return 0 === deepCompare(stateHash.bytes, hasher.digest());
};

const findLinkIndex = (branchNode, targetLinkHash) => (
	branchNode.links.findIndex(link => undefined !== link && 0 === deepCompare(targetLinkHash.bytes, link.bytes))
);

/**
 * Proves a patricia merkle hash.
 * @param {Hash256} encodedKey Encoded key of the state to prove.
 * @param {Hash256} valueToTest Expected hash of the state to prove.
 * @param {Array<TreeNode>} merklePath Merkle *node* path from root to leaf.
 * @param {Hash256} stateHash State hash from a block header.
 * @param {Array<Hash256>} subcacheMerkleRoots Sub cache merkle roots corresponding to the state hash.
 * @returns {number} Proof result code.
 */
const provePatriciaMerkle = (encodedKey, valueToTest, merklePath, stateHash, subcacheMerkleRoots) => {
	if (!checkStateHash(stateHash, subcacheMerkleRoots))
		return PatriciaMerkleProofResult.STATE_HASH_DOES_NOT_MATCH_ROOTS;

	const pathRootHash = merklePath[0].calculateHash();
	if (subcacheMerkleRoots.every(root => 0 !== deepCompare(pathRootHash.bytes, root.bytes)))
		return PatriciaMerkleProofResult.UNANCHORED_PATH_TREE;

	// positive proof must end with a leaf
	const isPositiveProof = 'value' in merklePath[merklePath.length - 1];
	if (isPositiveProof) {
		if (0 !== deepCompare(valueToTest.bytes, (/** @type LeafNode */ (merklePath[merklePath.length - 1])).value.bytes))
			return PatriciaMerkleProofResult.LEAF_VALUE_MISMATCH;
	}

	let childHash;
	let actualPath = '';
	for (let i = merklePath.length - 1; 0 <= i; --i) {
		const node = merklePath[i];
		const nodeHash = node.calculateHash();
		let formattedLinkIndex = '';
		if (childHash) {
			const linkIndex = findLinkIndex(node, childHash);
			if (-1 === linkIndex)
				return PatriciaMerkleProofResult.UNLINKED_NODE;

			formattedLinkIndex = uint8ToHex(new Uint8Array([linkIndex]))[1];
		}

		childHash = nodeHash;
		actualPath = `${formattedLinkIndex}${node.hexPath}${actualPath}`;
	}

	if (isPositiveProof) {
		// for positive proof, expected and calculated paths must match exactly
		return actualPath !== encodedKey.toString() ? PatriciaMerkleProofResult.PATH_MISMATCH : PatriciaMerkleProofResult.VALID_POSITIVE;
	}

	// for negative proof, expected path must start with calculated path and next nibble must be a dead end
	if (!encodedKey.toString().startsWith(actualPath))
		return PatriciaMerkleProofResult.PATH_MISMATCH;

	const nextNibble = getNibbleAt({ path: encodedKey.bytes, size: 2 * encodedKey.bytes.length }, actualPath.length);
	const nextNode = (/** @type BranchNode */ (merklePath[merklePath.length - 1])).links[nextNibble];
	return undefined !== nextNode ? PatriciaMerkleProofResult.INCONCLUSIVE : PatriciaMerkleProofResult.VALID_NEGATIVE;
};

// endregion

export {
	proveMerkle,
	deserializePatriciaTreeNodes,
	provePatriciaMerkle
};

// region type declarations

/**
 * Path in a Patricia merkle treee.
 * @class
 * @typedef {object} PatriciaTreePath
 * @property {Uint8Array} path Bytes composing the full path.
 * @property {number} size Length (in nibbles) of the path.
 */

/**
 * Represents part of a merkle tree proof.
 * @class
 * @typedef {object} MerklePart
 * @property {Hash256} hash Hash at this node.
 * @property {boolean} isLeft \c true if this is a left node; right otherwise.
 */

// endregion
